File: Top200/86-Partition.java
package Top200;

class Partition {
    class ListNode {
        int val;
        ListNode next;
        
        ListNode() {}
        
        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode partition(ListNode head, int x) {
        ListNode listOne = new ListNode(-1);
        ListNode listTwo = new ListNode(-1);
        ListNode dummy = new ListNode(-1);
        ListNode curr = head;
        ListNode currOne = listOne;
        ListNode currTwo = listTwo; 

        while (curr != null) {
            if (curr.val < x) {
                currOne.next = curr;
                curr = curr.next;
                currOne = currOne.next;
                currOne.next = null;
            } else {
                currTwo.next = curr;
                curr = curr.next;
                currTwo = currTwo.next;
                currTwo.next = null;
            }
        }

        currOne.next = listTwo.next;
        dummy.next = listOne.next;

        return dummy.next;
    }
}

File: Top200/86-Partition.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	l1 := &ListNode{Val: -1, Next: nil}
	l2 := &ListNode{Val: -1, Next: nil}
	l1Bak := l1
	l2Bak := l2
	nl := &ListNode{Val: -1, Next: nil}
	curr := head

	for curr != nil {
		if curr.Val < x {
			l1.Next = curr
			l1 = l1.Next
			curr = curr.Next
			l1.Next = nil
		} else {
			l2.Next = curr
			l2 = l2.Next
			curr = curr.Next
			l2.Next = nil
		}
	}

	l1.Next = l2Bak.Next
	nl.Next = l1Bak.Next

	return nl.Next
}
